crypto/Cargo.toml

```diff
@@ -34,10 +55,45 @@ parity-scale-codec = { workspace = true, features = ["derive", "full"] }
 serde = { workspace = true, features = ["derive"] }
 serde_with = { workspace = true, features = ["macros"] }
 hex = { workspace = true, features = ["alloc", "serde"] }
+# TODO: iroha_crypto no longer depends on openssl (did it ever?)
+# currently it's being used by iroha_client through attohttpc and iroha_cli through warp's tokeio-tunstenite
 openssl-sys = { version = "0.9.93", features = ["vendored"], optional = true }
```

Comment:
so why do we have this dependency if it doesn't depend on `openssl`?

Reply:
I think it's needed to force the `vendored` feature, which will make sure an openssl version bundled with the crate is built and used, instead of relying on system openssl. It also probably helps to ensure static linking is possible.

I haven't researched that much whether do we need it, but it's clear that even if we do, `iroha_crypto` is not the place.

`iroha_client` should probably be configurable by the user (you may want to use rustls, and `attohttpc` exposes features to do that)

Reply:
move it to `dev-dependencies`

Reply:
So, I did some experimentation and here's what I learned:

The `vendored` feature was added to make musl build of iroha work.

When building for musl, one cannot rely on the system openssl and has to build one themselves in order to statically link with it. So, the `vendored` feature was added to `iroha_crypto` to enable the `vendored` feature of `openssl-sys` which ultimately compiled the correct openssl with the correct compiler and linked it in.

But the thing is that `iroha_cli` (which is what gets built with musl in `Dockerfile`) does not depend on `openssl` at all. However other crates from the workspace (like `iroha_client` and `iroha_client_cli`) do. The whole workspace was getting build in the `Dockerfile`, so it failed.

I removed the `vendored` feature and modified the Dockerfile command to only build `iroha_cli`, so the build passes without this feature now. Ping @6r1d because they were the one dealing with Docker images AFAIR?

Reply:
Update, turns out the feature _is_ needed for the build of the docker image, as `iroha_client_cli` is bundled inside.

I've moved the `vendored` feature to `iroha_client_cli`, where it belongs

Reply:
why would `iroha_client_cli` need `openssl`?

Reply:
> However other crates from the workspace (like iroha_client and iroha_client_cli) do

yes, but only transitively through `iroha_crypto` crate

Reply:
`iroha_crypto` doesn't actually use any `openssl` at all, it's all pure rust!

What does use `openssl` is `native-tls`, which is in turn used by `attohttpc`, the http client used by `iroha-client` and `tokio-tungstenite` the websocket client used by `iroha-client`. The `native-tls` feature is even [explicitly enabled](https://github.com/hyperledger/iroha/blob/e9c6fe4f14dba2376c1ac1af85283b146c7005b1/client/Cargo.toml#L47) in `iroha_client`.

There is a common pattern now in rust libraries supporting HTTPS: they have two backends: `native-tls` (which works through openssl) and `rustls` (which is a pure-rust implementation based on `ring`).

We can ditch `native-tls` and use `rustls`, which brings much less linking trouble. Though, IMO, it only makes sense for binaries we ship; ideally the `iroha_client` should give the user a choice like all the other libraries ([1](https://github.com/elastic/elasticsearch-rs/blob/6c0bcd23eed35ad8492f69fd213bfcee52882156/elasticsearch/Cargo.toml#L25), [2](https://github.com/XAMPPRocky/octocrab/blob/551871a5766339f47119ee84686ce2dd4310e6ba/Cargo.toml#L73), [3](https://github.com/durch/rust-s3/blob/5105a1bff57827f88fb90840f03193665ea34d68/s3/Cargo.toml#L108), ...)

Reply:
Went ahead and did this in the latest commit. 

To reiterate: `iroha_crypto` does not use `openssl` (only for tests), but `openssl` is used for HTTPS support in `iroha_client`. This is not the only TLS implementation, `rustls` can be used for painless musl support.

This commit gives the user more freedom in choosing their TLS implementation by providing four `iroha_client` features: `tls-native`, `tls-native-vendored`, `tls-rustls-native-roots` and `tls-rustls-webpki-roots`, which mirror corresponding features of `attohttpc` and `tokio-tungstenite`.

`iroha_client_cli`, which is the only thing that uses `iroha_client` and gets built with musl libc, now uses the `tls-rustls-native-roots` feature, which does not need any tweaks to work with `musl`

Unlike previously, none of the TLS implementations are enabled by default, which is a breaking change. If this gets landed as-is, we'll have to communicate it to users and/or tech writers

Reply:
good job